Fix for #2423. request.data => body, params is used for querystrings

erikvullings opened the pull request:
## Description

- Replaced request.data with request.body.
- Added request.params to the signature table.
- Where request.data was used for querystring interpolation, I've used request.params.
- serialize uses body too
- replaced hard tabs with double spaces, so the markdown linter is happy.

## Motivation and Context

This is suppose to fix #2423, m.request documentation.
 
## How Has This Been Tested?

Not

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Documentation change

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have updated `docs/change-log.md`

osban commented:
In the table, `request.params` seems to have extra spaces compared to the other entries? Love the double spaces :)

spacejack commented:
Thanks for all that work Erik!

@isiahmeadows  I hate to mention it, but there is also jsonp: https://github.com/MithrilJS/mithril.js/blob/302b270fbd35f9fdac590e842b0d68911fe75d60/docs/jsonp.md
is `data` renamed `body` there as well?

isiahmeadows commented:
@spacejack It's `params` there, matching `m.request`. I apparently forgot to update the docs like I *thought* I did with that. Sorry about the mess...no idea how I missed that.

@erikvullings Just for sanity's sake, could you also check for all instances of `data` for `m.jsonp` and `m.request` in `docs/` and replace them as needed to `params` for URL parameters and `body` for request bodies?

shadowhand commented on docs/request.md:
```diff
@@ -25,12 +25,12 @@ Makes XHR (aka AJAX) requests, and returns a [promise](promise.md)
-	data: {id: 1, name: "test"}
+  method: "PUT",
+  url: "/api/v1/users/:id",
+  body: {id: 1, name: "test"}
```
Aren't the URL parameters are only replaced from `params` and not from `body`?

isiahmeadows replied:
Yep, this needs fixed.

erikvullings replied:
OK

shadowhand commented on docs/request.md:
```diff
@@ -44,8 +44,9 @@ Argument                  | Type                              | Required | Descr
-`options.url`             | `String`                          | Yes      | The [path name](paths.md) to send the request to, optionally interpolated with values from `options.data`.
-`options.data`            | `any`                             | No       | The data to be interpolated into the URL and serialized into the querystring (for GET requests) or body (for other types of requests).
+`options.url`             | `String`                          | Yes      | The [path name](paths.md) to send the request to, optionally interpolated with values from `options.params`.
+`options.params`            | `any`                           | No       | The data to be interpolated into the URL and serialized into the querystring (for GET requests).
```
Pretty sure this should be `Object` instead of `any`...?

erikvullings replied:
Changed

shadowhand commented on docs/request.md:
```diff
@@ -44,8 +44,9 @@ Argument                  | Type                              | Required | Descr
-`options.data`            | `any`                             | No       | The data to be interpolated into the URL and serialized into the querystring (for GET requests) or body (for other types of requests).
+`options.url`             | `String`                          | Yes      | The [path name](paths.md) to send the request to, optionally interpolated with values from `options.params`.
+`options.params`            | `any`                           | No       | The data to be interpolated into the URL and serialized into the querystring (for GET requests).
+`options.body`            | `any`                             | No       | The data to be serialized into the body (for other types of requests).
```
Ditto here, `Object` instead of `any` ?

erikvullings replied:
OK

shadowhand commented on docs/request.md:
```diff
@@ -55,10 +56,10 @@ Argument                  | Type                              | Required | Descr
 `options.deserialize`     | `any = Function(any)`          | No       | A deserialization method to be applied to the `xhr.response` or normalized `xhr.responseText`. Defaults to the [identity function](https://en.wikipedia.org/wiki/Identity_function). If `extract` is defined, `deserialize` will be skipped.
 `options.extract`         | `any = Function(xhr, options)`    | No       | A hook to specify how the XMLHttpRequest response should be read. Useful for processing response data, reading headers and cookies. By default this is a function that returns `options.deserialize(parsedResponse)`, throwing an exception when the server response status code indicates an error or when the response is syntacti ...
-`options.useBody`         | `Boolean`                         | No       | Force the use of the HTTP body section for `data` in `GET` requests when set to `true`, or the use of querystring for other HTTP methods when set to `false`. Defaults to `false` for `GET` requests and `true` for other methods.
+`options.useBody`         | `Boolean`                         | No       | Force the use of the HTTP body section for `body` in `GET` requests when set to `true`, or the use of querystring for other HTTP methods when set to `false`. Defaults to `false` for `GET` requests and `true` for other methods.
```
This line should be removed, as `useBody` was dropped in #2361.

erikvullings replied:
Removed

shadowhand commented on docs/request.md:
```diff
@@ -82,11 +83,11 @@ The `m.request` utility is a thin wrapper around [`XMLHttpRequest`](https://deve
-	method: "GET",
-	url: "/api/v1/users",
+  method: "GET",
+  url: "/api/v1/users",
```
Why are whitespace changes included in this PR?

erikvullings replied:
My bad. I use a markdown linter, and it replaced them all...

shadowhand commented on docs/request.md:
```diff
@@ -272,27 +273,27 @@ Next, you need to create a [`FormData`](https://developer.mozilla.org/en/docs/We
-	var data = new FormData()
-	data.append("myfile", file)
+  var data = new FormData()
+  data.append("myfile", file)
```
I think it would be better if this used `body` instead of `data` to make it clear what option should be used.

isiahmeadows replied:
Agreed.

erikvullings replied:
OK.

shadowhand commented on docs/request.md:
```diff
@@ -272,27 +273,27 @@ Next, you need to create a [`FormData`](https://developer.mozilla.org/en/docs/We
+  m.request({
+    method: "POST",
+    url: "/api/v1/upload",
+    body: data,
```
Similarly here.

erikvullings replied:
OK

shadowhand commented on docs/request.md:
````diff
@@ -202,23 +203,23 @@ Request URLs may contain interpolations:
 })
 ```
 
 In the code above, `:id` is populated with the data from the `{id: 123}` object, and the request becomes `GET /api/v1/users/123`.
````
```suggestion
In the code above, `:id` is populated with the data from the `params` object, and the request becomes `GET /api/v1/users/123`.
```

isiahmeadows replied:
This line needs fixed (and the above suggestion would work).

(Including this purely for the sake of review.)

erikvullings replied:
OK

shadowhand commented on docs/request.md:
```diff
@@ -55,10 +56,10 @@ Argument                  | Type                              | Required | Descr
 `options.deserialize`     | `any = Function(any)`          | No       | A deserialization method to be applied to the `xhr.response` or normalized `xhr.responseText`. Defaults to the [identity function](https://en.wikipedia.org/wiki/Identity_function). If `extract` is defined, `deserialize` will be skipped.
 `options.extract`         | `any = Function(xhr, options)`    | No       | A hook to specify how the XMLHttpRequest response should be read. Useful for processing response data, reading headers and cookies. By default this is a function that returns `options.deserialize(parsedResponse)`, throwing an exception when the server response status code indicates an error or when the response is syntacti ...
-`options.useBody`         | `Boolean`                         | No       | Force the use of the HTTP body section for `data` in `GET` requests when set to `true`, or the use of querystring for other HTTP methods when set to `false`. Defaults to `false` for `GET` requests and `true` for other methods.
+`options.useBody`         | `Boolean`                         | No       | Force the use of the HTTP body section for `body` in `GET` requests when set to `true`, or the use of querystring for other HTTP methods when set to `false`. Defaults to `false` for `GET` requests and `true` for other methods.
```
```suggestion
```

isiahmeadows replied:
This line needs fixed (and the above suggestion would work).

(Including this purely for the sake of review.)

erikvullings replied:
OK

shadowhand commented on docs/request.md:
```diff
@@ -44,8 +44,9 @@ Argument                  | Type                              | Required | Descr
-`options.url`             | `String`                          | Yes      | The [path name](paths.md) to send the request to, optionally interpolated with values from `options.data`.
-`options.data`            | `any`                             | No       | The data to be interpolated into the URL and serialized into the querystring (for GET requests) or body (for other types of requests).
+`options.url`             | `String`                          | Yes      | The [path name](paths.md) to send the request to, optionally interpolated with values from `options.params`.
+`options.params`            | `any`                           | No       | The data to be interpolated into the URL and serialized into the querystring (for GET requests).
```
```suggestion
`options.params`            | `Object`                        | No       | The data to be interpolated into the URL and/or serialized into the query string.
```

isiahmeadows replied:
This line needs fixed (and the above suggestion would work).

(Including this purely for the sake of review.)

erikvullings replied:
Copied

shadowhand commented on docs/request.md:
```diff
@@ -44,8 +44,9 @@ Argument                  | Type                              | Required | Descr
-`options.data`            | `any`                             | No       | The data to be interpolated into the URL and serialized into the querystring (for GET requests) or body (for other types of requests).
+`options.url`             | `String`                          | Yes      | The [path name](paths.md) to send the request to, optionally interpolated with values from `options.params`.
+`options.params`            | `any`                           | No       | The data to be interpolated into the URL and serialized into the querystring (for GET requests).
+`options.body`            | `any`                             | No       | The data to be serialized into the body (for other types of requests).
```
```suggestion
`options.body`            | `Object`                          | No       | The data to be serialized into the body.
```

isiahmeadows replied:
This line needs fixed (and the above suggestion would work).

(Including this purely for the sake of review.)

erikvullings replied:
OK

shadowhand commented on docs/request.md:
````diff
@@ -202,23 +203,23 @@ Request URLs may contain interpolations:
+  params: {id: 123},
 })
 ```
 
````
```suggestion

In the code above, the request becomes `GET /api/v1/users/foo:bar?id=123`.
```

isiahmeadows replied:
This particular suggestion is optional.

(Including this purely for the sake of review.)

erikvullings replied:
I've included it - I believe it is important, so users will know that if their `params` contains a lot of properties, they will all be added to the URL.

erikvullings commented:
> @erikvullings Just for sanity's sake, could you also check for all instances of `data` for `m.jsonp` and `m.request` in `docs/` and replace them as needed to `params` for URL parameters and `body` for request bodies?

OK, will do.

shadowhand commented on docs/request.md:
```diff
@@ -27,7 +27,7 @@ Makes XHR (aka AJAX) requests, and returns a [promise](promise.md)
 	method: "PUT",
 	url: "/api/v1/users/:id",
-	data: {id: 1, name: "test"}
+	params: {id: 1, name: "test"}
```
```suggestion
	params: {id: 1},
	body: {name: "test"}
```

isiahmeadows replied:
Why?

shadowhand replied:
The request method is `PUT`, so it needs a body. The current example will produce:

```
PUT /api/v1/users/1
```

with an empty body. 😨

isiahmeadows replied:
It works well enough for the examples, and I've done put-like operations using the query string only. Your app obviously will be more complicated, and this is purely just for show, nothing more.

shadowhand replied:
This is a fundamental change in this example. Previously, it would have made an `application/json` request without a query string.

If I make a PR to fix this, will you accept it?

isiahmeadows replied:
I'll consider it as long as you include this context, but keep in mind `id` has to be in `params` for it to be interpolated.

shadowhand replied:
https://github.com/MithrilJS/mithril.js/pull/2430

erikvullings commented:
I've followed your recommendations and changed the pull request accordingly. I've also had a look at the other documents that were using `data`, and replaced it with `body` or `params` accordingly.

isiahmeadows commented on docs/request.md:
```diff
@@ -497,7 +497,7 @@ Mithril's `m.request` uses `XMLHttpRequest` instead of `fetch()` for a number of
 Currently, due to lack of browser support, `fetch()` typically requires a [polyfill](https://github.com/github/fetch), which is over 11kb uncompressed - nearly three times larger than Mithril's XHR module.
 
-Despite being much smaller, Mithril's XHR module supports many important and not-so-trivial-to-implement features like [URL interpolation](#dynamic-urls), querystring serialization and [JSON-P requests](jsonp.md), in addition to its ability to integrate seamlessly to Mithril's autoredrawing subsystem. The `fetch` polyfill does not support any of those, and requires extra libraries and boilerplate ...
+Despite being much smaller, Mithril's XHR module supports many important and not-so-trivial-to-implement features like [URL interpolation](#dynamic-urls), querystring serialization and [JSON-P requests](jsonp.md), in addition to its ability to integrate seamlessly to Mithril's auto-redrawing subsystem. The `fetch` polyfill does not support any of those, and requires extra libraries and boilerplat ...
```
I'm reverting this change since the docs use both interchangeably.
